include/pybind11/numpy.h

```diff
@@ -810,9 +870,7 @@ class array : public buffer {
     }
 
     /// Byte size of a single element
-    ssize_t itemsize() const {
-        return detail::array_descriptor_proxy(detail::array_proxy(m_ptr)->descr)->elsize;
-    }
+    ssize_t itemsize() const { return dtype().itemsize(); }
```

Comment:
So, to be more clear, there are two downstream changes:
1. The itemsize/alignment of the dtype (not array!) is changed to `ssize_t` and the flags to `uint64`.
2. The change here if users use the `PyArray_Proxy`.
 
   If the user uses `array_proxy(arr)->descr` (can they reasonably, as this is    "detail"?), then the returned `PyArray_Proxy->descr` field loses access to fields    related to the structured dtypes (that is `fields` and `names` really).

   It also loses access to `->elsize` and `->alignment`, but for those it would be odd to not use the existing functionality.

I can't decide whether these warrant a `numpy2.h`.  To me it the vast majority of code won't hit them, but if you like strict principles you can argue compilation could break due to it.

OTOH, I don't think the struct change here is even visible unless you use `detail::` API, and code that does that should probably accept having to adapt in exceptional circumstances (which the upcoming NumPy 2 release clearly is).

(I am not counting the getparams function removal: it has been spewing warnings/errors for years.)

Reply:
I think this is fine, as users should not be using detail::